hazelcast/src/test/java/com/hazelcast/client/usercodedeployment/ClientUserCodeDeploymentTest.java

```diff
@@ -158,6 +161,46 @@ public void testWithMultipleNodes_clientReconnectsToNewNode() throws Interrupted
         assertCodeDeploymentWorking(client, new IncrementingEntryProcessor());
     }
 
+    @Test
+    public void testWithSplitBrain_clientReconnectsToOtherHalf() {
+        ClientConfig clientConfig = createClientConfig();
+        Config config = createNodeConfig();
+
+        HazelcastInstance instance1 = factory.newHazelcastInstance(config);
+        HazelcastInstance instance2 = factory.newHazelcastInstance(config);
+
+        assertClusterSizeEventually(2, instance1, instance2);
```

Comment:
After client connects to the cluster initially here, it deploys class definitions to the cluster because client state is sent in `HazelcastClientInstanceImpl.start()`. So, in order to assert that client state is sent again to the new cluster without cluster id changing, should not you somehow remove the class defitinion from the to-be-reconnected member? 

Also, in addition to testing end-to-end behaviours, (i.e testing compact schemas and user code deployment) you can add a unit test that asserts `sendStateToCluster` is called in split brain reconnection scenario. I think it would directly test your change instead of testing implicitly.

Reply:
No, the client connects only one-half of the cluster (See the member list size assertion). 

How can you write a unit test that simulates the split-brain? Sure, it might be possible but I don't think the benefits would worth the huge effort needed to do that.